airbyte-cdk/java/airbyte-cdk/db-sources/src/testFixtures/java/io/airbyte/cdk/testutils/ContainerFactory.java

```diff
@@ -22,97 +24,106 @@
  * ContainerFactory is the companion interface to {@link TestDatabase} for providing it with
  * suitable testcontainer instances.
  */
-public interface ContainerFactory<C extends JdbcDatabaseContainer<?>> {
+public abstract class ContainerFactory<C extends JdbcDatabaseContainer<?>> {
 
-  /**
-   * Creates a new, unshared testcontainer instance. This usually wraps the default constructor for
-   * the testcontainer type.
-   */
-  C createNewContainer(DockerImageName imageName);
+  static private final Logger LOGGER = LoggerFactory.getLogger(ContainerFactory.class);
 
-  /**
-   * Returns the class object of the testcontainer.
-   */
-  Class<?> getContainerClass();
+  private record ContainerKey(Class<? extends ContainerFactory> clazz, DockerImageName imageName, List<String> methods) {};
 
-  /**
-   * Returns a shared instance of the testcontainer.
-   */
-  default C shared(String imageName, String... methods) {
-    final String mapKey = Stream.concat(
-        Stream.of(imageName, this.getClass().getCanonicalName()),
-        Stream.of(methods))
-        .collect(Collectors.joining("+"));
-    return Singleton.getOrCreate(mapKey, this);
-  }
+  private static class ContainerOrException {
 
-  /**
-   * This class is exclusively used by {@link #shared(String, String...)}. It wraps a specific shared
-   * testcontainer instance, which is created exactly once.
-   */
-  class Singleton<C extends JdbcDatabaseContainer<?>> {
-
-    static private final Logger LOGGER = LoggerFactory.getLogger(Singleton.class);
-    static private final ConcurrentHashMap<String, Singleton<?>> LAZY = new ConcurrentHashMap<>();
+    private final Supplier<GenericContainer<?>> containerSupplier;
+    private volatile RuntimeException _exception = null;
+    private volatile GenericContainer<?> _container = null;
 
-    @SuppressWarnings("unchecked")
-    static private <C extends JdbcDatabaseContainer<?>> C getOrCreate(String mapKey, ContainerFactory<C> factory) {
-      final Singleton<?> singleton = LAZY.computeIfAbsent(mapKey, Singleton<C>::new);
-      return ((Singleton<C>) singleton).getOrCreate(factory);
+    ContainerOrException(Supplier<GenericContainer<?>> containerSupplier) {
+      this.containerSupplier = containerSupplier;
```

Comment:
add commentary somewhere in this cass explaining why we bother going through an extra level of indirection with a `Supplier` instead of eagerly constructing the container